text/0057-general-multiplication.md

```diff
@@ -0,0 +1,92 @@
+# General multiplication in Cedar
+
+## Related issues and PRs
+
+- Reference Issues: related to [RFC 50]
+- Implementation PR(s): [cedar#702]
+
+## Timeline
+
+- Started: 2024-03-07
+
+## Summary
+
+Allow multiplication of arbitrary expressions (that evaluate to Long), not just
+multiplication of an expression by a constant.
+
+## Basic example
+
+```
+permit(principal, action, resource) when {
+    context.foo * principal.bar >= 100
+};
+```
+
+## Motivation
+
+Today, Cedar only supports multiplication of two operands if (at least) one
+operand is a constant.
+In the more general n-ary (chained) multiplication case, Cedar requires that at
+most one operand is a non-constant, but allows the constants and non-constant to
+appear in any order.
+
+- This creates some confusion for Cedar users. No other popular programming
+  language has this restriction on multiplication.
+- This leads to bugs and confusing behavior. To enforce this restriction,
+  Cedar's internal AST node for multiplication only supports multiplying an
+  expression by a constant. But this requires the Cedar parser to re-associates
+  (and re-order) operands in a chained multiplication, which leads to confusion
+  and problems as described in [RFC 50].
+- This limits the power of Cedar by preventing users from writing policies
+  involving multiplication where neither operand is known at policy-authoring
+  time.
+
+This RFC proposes to relax all of these restrictions and allow multiplication
+of arbitrary expressions (that evaluate to Long).
+In other words, multiplication will be treated just like addition: `*` will be
+valid in all positions and situations where `+` is valid today.
+
+## Detailed design
+
+This will actually simplify the evaluator code, validator code, and internal AST.
+Multiplication will become an ordinary `BinaryOp`, and no longer need a special
+AST node.
+The evaluator and validator can handle multiplication in the same way as
+addition and in the same code paths.
+
+This RFC does not introduce any new categories of errors, and does not result in
+changed behavior for any policies that are valid today.
```

Comment:
Are there any expressions for which the previous semantics will not overflow, but the new one will? One half-baked idea based on #50: I would expect an expression like `CONST * CONST * content.value` to overflow for large `CONST` and `content.value=0` under the proposed semantics. But under the old semantics, during parsing this expression is converted to `content.value * CONST * CONST` which cannot overflow.

Should we consider this a behavior change?

Reply:
Unfortunately, your example is correct.  I will update the language here.

Reply:
Revised.